packages/rum-recorder/src/domain/rrweb/mutationObserver.ts

```diff
@@ -0,0 +1,274 @@
+import { monitor } from '@datadog/browser-core'
+import {
+  getSerializedNodeId,
+  hasSerializedNode,
+  transformAttribute,
+  NodeWithSerializedNode,
+  nodeIsIgnored,
+  serializeNodeWithId,
+  IdNodeMap,
+  nodeOrAncestorsIsIgnored,
+} from '../rrweb-snapshot'
+import { nodeOrAncestorsShouldBeHidden } from '../privacy'
+import {
+  AddedNodeMutation,
+  AttributeMutation,
+  AttributesMutationRecord,
+  CharacterDataMutationRecord,
+  ChildListMutationRecord,
+  MutationCallBack,
+  MutationRecord,
+  RemovedNodeMutation,
+  TextMutation,
+} from './types'
+import { forEach } from './utils'
+import { MutationController } from './mutation'
+
+type WithSerializedTarget<T> = T & { target: NodeWithSerializedNode }
+
+/**
+ * Buffers and aggregate mutations generated by a MutationObserver into MutationPayload
+ */
+export function startMutationObserver(controller: MutationController, mutationCallback: MutationCallBack) {
+  const observer = new MutationObserver(monitor((mutations) => processMutations(mutations, mutationCallback)))
+  observer.observe(document, {
+    attributeOldValue: true,
+    attributes: true,
+    characterData: true,
+    characterDataOldValue: true,
+    childList: true,
+    subtree: true,
+  })
+  controller.onFlush(() => processMutations(observer.takeRecords(), mutationCallback))
+
+  return {
+    stop: () => {
+      observer.disconnect()
+    },
+  }
+}
+
+function processMutations(mutations: MutationRecord[], mutationCallback: MutationCallBack) {
+  // Discard any mutation with a 'target' node that:
+  // * isn't injected in the current document or isn't known/serialized yet: those nodes are likely
+  // part of a mutation occurring in a parent Node
+  // * should be hidden or ignored
+  const filteredMutations = mutations.filter(
+    (mutation): mutation is WithSerializedTarget<MutationRecord> =>
+      document.contains(mutation.target) &&
+      hasSerializedNode(mutation.target) &&
+      !nodeOrAncestorsIsIgnored(mutation.target) &&
+      !nodeOrAncestorsShouldBeHidden(mutation.target)
```

Comment:
Yes I agree, this is still something I have in mind: optimizing this loop or using the 'cache' optimization could still be usefull here. In practice, I observed that this whole filter loop takes less than 1ms even when hundreds of mutations occurs.

`document.contains` is **fast**. It is not O(N), and the browser probably uses internal tricks to avoid walking on all parents. See for yourself, run this in a busy page:

```js
{ const list = document.querySelectorAll('*'); const start = performance.now(); list.forEach(e => { while (e) { e = e.parentNode };  }); console.log(performance.now() - start) }

{ const list = document.querySelectorAll('*'); const start = performance.now(); list.forEach(e => document.contains(e)); console.log(performance.now() - start) }
```